compiler/optimizations/gpuTransforms.cpp

```diff
@@ -1271,7 +1290,9 @@ static void doGpuTransforms() {
 
   // Outline all eligible loops; cleanup CPU bound loops
   forv_Vec(FnSymbol*, fn, gFnSymbols) {
-    if(fGpuSpecialization && !fn->hasFlag(FLAG_GPU_SPECIALIZATION)) {
+    if(fGpuSpecialization && !fn->hasFlag(FLAG_GPU_SPECIALIZATION)
+     && recklesslyDisregardSpecializationLimits)
+    {
```

Comment:
This doesn't meet "wait until the 3rd use" for refactoring criteria, but it took me a while to understand how the flag was used (I was assuming it to be `true` initially for some reason, maybe the name isn't blunt enough).

Can we have a `static bool isFnSpecialized(FnSymbol* fn)` helper (or its equivalent private class function if that makes more sense) and use it here and above? I'd find that definitely more self-documenting.

Reply:
Sure I added a function to check the flag and pulled the conditions out into named variables:

```C++
      bool canAssumeFnWillRunOnCpu = fGpuSpecialization &&
                                     !isFnGpuSpecialized(fn) &&
                                     assumeNonGpuSpecFnsAreOnCpu;
      if(canAssumeFnWillRunOnCpu)
```

```C++
FnSymbol *fnContainingLoop = gpuLoop.loop()->getFunction();
bool canAssumeFnWillRunOnGpu =
  fGpuSpecialization && (assumeNonGpuSpecFnsAreOnCpu || 
  isFnGpuSpecialized(fnContainingLoop));
if(canAssumeFnWillRunOnGpu)
```

Reply:
Oh wait, I actually was thinking the `isFnGpuSpecialized` would also check for `assumeNonGpuSpecFnsAreOnCpu` in fact it could even check for `fGpuSpecialization`. If that doesn't make sense to you, let's not dwell too much on it. `isFnGpuSpecialized` as-is is improvement -- at least it reduces the primitive's appearance in the code.